File: Range-Sum-Query---Immutable.js
var NumArray = function(nums) 
{
    this.p = [0];
    for (let i = 0; i < nums.length; i++) 
    {
        this.p[i + 1] = this.p[i] + nums[i];
    }
};

NumArray.prototype.sumRange = function(left, right) 
{
    return this.p[right + 1] - this.p[left];
};

File: Range-Sum-Query---Immutable.java
class NumArray 
{
    private int[] pre;

    public NumArray(int[] nums) 
    {
        pre = new int[nums.length + 1];
        pre[0] = 0;
        for (int i = 0; i < nums.length; i++)
         {
            pre[i + 1] = pre[i] + nums[i];
        }
    }

    public int sumRange(int left, int right) 
    {
        return pre[right + 1] - pre[left];
    }
}
